### ISSUE TITLE
Call into view write path when applying streaming mutations

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#3275]
*Installation details*
Scylla version (or git commit hash): master

During streaming, there are cases when we should invoke the view write path. In particular, if a view has not yet finished building and we're bootstrapping a new node.

We should check whether the view is built across all base replicas responsible for the range in question. We could query just the base replica that's currently paired with view replica we will be paired with, but we may not have a ring snapshot at that point, and it may not be stable if we support multiple range movements. Better to be conservative and query all of them base nodes.

**Comment by duarten** [comment]
The fragmented "big mutations" complicate things a bit, as we may only be able to process those mutations once we flush a particular memtable.

**Comment by avikivity** [comment]
Those are going away. You'll get an incoming stream of mutation_fragments.

If you want to process only after the stream was committed, then we can do that too. The stream will bypass memtables and generate its own sstables, which you can then read from.

**Comment by slivne** [comment]
2 questions/comments

- Duarte:
 - It is clear to me why during repair/rebuild etc we need to call into the
view path during streaming.
 - Why do we need during bootstrap to call into the view path for the
streaming information ?


- Avi
 - Views require that we read the old value and than compute the delta - I
am not sure we can wait and read from the disk after the streaming has
ended - we will not be able to read the old value. We need to find a
solution to allow views ability to access the the "old" values.

**Comment by duarten** [comment]
Maybe we don't, but my reasoning is as follows. The view building process for the range we're getting may not have yet finished, and may not finish before the range migration is complete and the old node is cleaned up. Now, I pin (for other reasons too) the set of sstables we're using for view building, so we should be ok if it's just a range movement. But if the source node is going to be decommissioned, or crashes and is lost, then we won't complete the view building. I think we should be conservative.

**Comment by slivne** [comment]
On Mon, Mar 12, 2018 at 11:59 AM, Duarte Nunes <notifications@github.com>
wrote:

> Maybe we don't, but my reasoning is as follows. The view building process
> for the range we're getting may not have yet finished, and may not finish
> before the range migration is complete and the old node is cleaned up. Now,
> I pin (for other reasons too) the set of sstables we're using for view
> building, so we should be ok if it's just a range movement.
>



> But if the source node is going to be decommissioned, or crashes and is
> lost, then we won't complete the view building. I think we should be
> conservative.
>
AFAIK the source node has to be there or else the bootstrap will not work -
so you can not decomission the node before the node has transferred all the
data.

I don't think we can have the bootstrap node doing view updates - it may
not have the previous (old) data - its view update commands may be
"incorrect'

- nadav/duarte - how does caddanra handle this ? does the bootsrtap node do
view updates ?

**Comment by duarten** [comment]
Yes, the node transfers all the data, and then is decommissioned. This doesn't mean it will have finished building all the views of the base table being transferred, specially if a view is added right before decommissioning.

Yes, they handle this and invoke the view write path on bootstrap, although they don't check the view_build_status distributed table, which is a bug (they have a patch pending for that, but hasn't been applied; I even reviewed it, but the author never got back to it).

Regarding old data, we don't really compute the delta, but rather the whole new row. If we receive mutations out-of-order, this means receiving a mutation with a timestamp lower than what already exists in the base table. This happen naturally (clock skew, clients setting their own timestamp, etc.), so the view building code should already handle this and not send anything. I have to check that this is indeed the case (and to be more precise, we may send something, and that is the columns that are not superseded by the most recent, existing update).

Also, even if we fail to cover all corner case, when we have proper view repair, we'll be alright :)

Building view updates from a stream of `mutation_fragments` isn't great and may cause a lot of churn, so better to build from the sstable I guess.

**Comment by avikivity** [comment]
How is the sstable better? it's the same stream of mutation_fragments, given form.

**Comment by duarten** [comment]
@avikivity For every incoming mutation_fragment we have to do a read-before-write of the row(s) and send out view updates, possibly resulting in a lot of churn (there can be a thousand rows setting values for a column that's included in the view or something). If the generated sstable is in the set we're reading from, then those rows will be coalesced and we'll only send out the minimum set of view updates.

**Comment by avikivity** [comment]
There is a 1:1 correspondence between the mutation fragments and the data in the sstable.

The only difference is that an RMW will not see data being streamed. So maybe the only option is to wait until the streaming is done.

**Comment by duarten** [comment]
Ah, I guess the incoming mutations will all be in the most compact form, so reading from them or from the sstable results in the same. 

Then, since each row is in its canonical form, we don't need to consider the other fragments in the RBW cycle and can indeed apply the view write path for each fragment.

**Comment by avikivity** [comment]
They are in canonical form, but only contains the sending node's data. I guess that's sufficient.

For streaming, we might combine the streams from different replicas of a vnode before writing them, so:

    [[existing replica sstable readers]] -> [combined reader] -> [send] -> [receive] -> combine -> you -> sstable writer

Each level of brackets indicates a many:1 relationship, one from multiple sstables per range, one from multiple replicas per vnode.

/cc @asias

**Comment by duarten** [comment]
> @psarna / @duarten - can you please detail in which cases streaming should call into view building write path (and which not) ?
> 
> * bootstrap
> * decommission
> * removenode
> * rebuild
> * repair

We need to call into the view building path for repair (this is because a base replica missed an updated) and for all others if the view has not finished building.

> 
> Instead of calling into the view building write path
> 
> * we could transfer what is "left" of building the view to the new owning nodes (anyhow, we need to stop the background build and not continue after the ownership of ranges has been completed).

We can't stop the background build, because we need to build other ranges as well. Also, we would only cancel when a cleanup is issued, because otherwise the user might not complete a bootstrap, for example.

I don't think we should complicate matters, we should just call into the view build write path.

> * we could stop the view building while streaming is progressing and resume it after streaming has ended for the ranges that are left (on the new "owning" node).

This will require changes to the view building code and more system tables for storing which ranges are left. 

> 
> The direction of calling into the view building write path while streaming is happening is "technically safe" - we will call into the view building write path twice once on the "old" node and once on the "new" node - yet its an overhead that maybe very high. We can also "pin down" the streamed sstables if that helps in some manner.

Is this really an issue? How often are there range movements while a view is still being built?

> 
> I think that for in flight writes (during bootstrap as an example) we call into the view write path today to make sure that views are updated.

Yeah, we consider the pending endpoints.

> 
> For streaming maybe a different solution should be employed.

**Comment by nyh** [comment]
On Wed, Sep 12, 2018 at 4:37 PM, Duarte Nunes <notifications@github.com>
wrote:

> @psarna <https://github.com/psarna> / @duarten
> <https://github.com/duarten> - can you please detail in which cases
> streaming should call into view building write path (and which not) ?
>
>    - bootstrap
>    - decommission
>    - removenode
>    - rebuild
>    - repair
>
> We need to call into the view building path for repair (this is because a
> base replica missed an updated) and for all others if the view has not
> finished building.
>
I think there's some confusion here given the title of this issue, whether
we're really talking about view *building* (i.e., the background process
which happens when we add a view to a base table with existing data) or
ordinary view writes. The comments seem to mix the two things, sometimes
even in the same sentence...

Regardless of view *building*, repair is a special issue which we indeed
need to handle (and I think we should have a separate issue on it. I was
sure we had, but can't find one.)
The question is what we expect to happen if one starts a repair only on the
base table - should it repair *just* the base table, or does it also do
something to the view table if it finds discrepancies?

Cassandra compatibility requires the latter:
https://docs.datastax.com/en/cql/3.3/cql/cql_using/knownLimitationsMV.html
states that:
*"Full repairs on base tables must go through the write path to generate
view updates".*

So if we want to be compatible with that, repair indeed needs to go through
the view write path. I wouldn't call it the "view building" write path
though - it's the full, normal, write path, which includes the view updates.

Instead of calling into the view building write path
>
>    - we could transfer what is "left" of building the view to the new
>    owning nodes (anyhow, we need to stop the background build and not continue
>    after the ownership of ranges has been completed).
>
>
This part of the discussion (which I don't understand) seems to talk about
a completely different issue - range movements (adding or deleting a node).
But what is missing here?
If I understand correctly (but @duarte, please correct me if I'm wrong),
the distributed view building code already handles the case of range
movements - what has already been built will be streamed, unbuilt ranges
will be built on the new view replica as well because the view building
code also takes into account "pending nodes" (at least, I hope it does!).
New writes that happen during the build also work properly, because of the
"pending nodes" support (this I know exists because I wrote it at some
point, and then Duarte improved on it).
All in all, @duarte, I don't understand the problem that this issue claims
exists.

**Comment by duarten** [comment]
On Wed, 12 Sep 2018 at 17:07, nyh <notifications@github.com> wrote:

> On Wed, Sep 12, 2018 at 4:37 PM, Duarte Nunes <notifications@github.com>
> wrote:
>
> > @psarna <https://github.com/psarna> / @duarten
> > <https://github.com/duarten> - can you please detail in which cases
> > streaming should call into view building write path (and which not) ?
> >
> > - bootstrap
> > - decommission
> > - removenode
> > - rebuild
> > - repair
> >
> > We need to call into the view building path for repair (this is because a
> > base replica missed an updated) and for all others if the view has not
> > finished building.
> >
> I think there's some confusion here given the title of this issue, whether
> we're really talking about view *building* (i.e., the background process
> which happens when we add a view to a base table with existing data) or
> ordinary view writes.


The issue is about streaming. It also predates the view_builder and by the
“building” in the title I mean the incremental, normal view write path.
I’ll update it to avoid confusion.


> The comments seem to mix the two things, sometimes
> even in the same sentence...


My comment doesn't mix anything, it merely states the cases which require
us to call into the view write path for streaming, some of which are
dependent on the view building status.


>
> Regardless of view *building*, repair is a special issue which we indeed
> need to handle (and I think we should have a separate issue on it. I was
> sure we had, but can't find one.)
> The question is what we expect to happen if one starts a repair only on the
> base table - should it repair *just* the base table, or does it also do
> something to the view table if it finds discrepancies?


A base repair means a replica missed some mutations, which of course need
to go through the view write path. A paired base<->view repair wouldn't
catch these inconsistencies.


>
> Cassandra compatibility requires the latter:
> https://docs.datastax.com/en/cql/3.3/cql/cql_using/knownLimitationsMV.html
> states that:
> *"Full repairs on base tables must go through the write path to generate
> view updates".*
>
> So if we want to be compatible with that, repair indeed needs to go through
> the view write path. I wouldn't call it the "view building" write path
> though - it's the full, normal, write path, which includes the view
> updates.


No, it's just the view write path. We don't need to apply the base
mutations, as those are written directly to an sstable.


>
> Instead of calling into the view building write path
> >
> > - we could transfer what is "left" of building the view to the new
> > owning nodes (anyhow, we need to stop the background build and not
> continue
> > after the ownership of ranges has been completed).
> >
> >
> This part of the discussion (which I don't understand) seems to talk about
> a completely different issue - range movements (adding or deleting a node).


Range movements are precisely what this issue is about, aside from base
repair.


> But what is missing here?
> If I understand correctly (but @duarte, please correct me if I'm wrong),
> the distributed view building code already handles the case of range
> movements - what has already been built will be streamed, unbuilt ranges
> will be built on the new view replica as well because the view building
>

View building happens on base replicas. But no, there's no synchronization
between a node learning about a view schema (for which it has no data),
view building, and range movements. That's what this issue is meant to
address.


> code also takes into account "pending nodes" (at least, I hope it does!).
> New writes that happen during the build also work properly, because of the
> "pending nodes" support (this I know exists because I wrote it at some
> point, and then Duarte improved on it).
> All in all, @duarte, I don't understand the problem that this issue claims
> exists.


Well, one problem you understand, which is repair. The other problem is
view building. If node A is building view V, but then its ranges R are
moved from A to B, there's no guarantee those ranges have been built.

**Comment by duarten** [comment]
Another thing to pay attention to is, for range movements, whether the target base replica will be paired with the same view replica as the source base replica. I think they will, but haven't done the exercise of proofing it for the `NetworkTopologyStrategy`.

**Comment by nyh** [comment]
It's very difficult to coherently discuss the many different sub-issues
involved in this issue, but I'll try:

On Wed, Sep 12, 2018 at 7:33 PM, Duarte Nunes <notifications@github.com>
wrote:

>
> >
> > Cassandra compatibility requires the latter:
> > https://docs.datastax.com/en/cql/3.3/cql/cql_using/
> knownLimitationsMV.html
> > states that:
> > *"Full repairs on base tables must go through the write path to generate
> > view updates".*
> >
> > So if we want to be compatible with that, repair indeed needs to go
> through
> > the view write path. I wouldn't call it the "view building" write path
> > though - it's the full, normal, write path, which includes the view
> > updates.
>
>
> No, it's just the view write path. We don't need to apply the base
> mutations, as those are written directly to an sstable.
>

I think when there are views, it is not safe to write directly to sstables
and
*also* send view updates to the paired view replica. One of the problems
I can think of with concurrent updates (in parallel with the streaming,
someone
is modifying the same partitions/rows). For concurrent updates, as you
surely
remember, we need to use a row/partition lock for the RMW. If we write to
sstables
but not read from them yet, or if we don't properly lock rows for the right
duration,
we can end up breaking concurrent updates. Unless I'm missing something?

So I think when views are involved, we need to pass throught the normal
write code - writing to memtables, sstables, views, RMW, lock - the works -
and can't optimize it. Again, unless I'm missing something. Maybe there's
another way, like keeping all the rows we write to the sstable locked, but
that doesn't look pretty either...

If I remember correctly, Cassandra does this as well - when there are no
views,
it streams sstables directly (as files), but when there are views, it
streams
mutations and applies them using the full, normal, write path (at least,
that's what I vaguely remember, I don't have the code in front of me right
now).

**Comment by duarten** [comment]
> I think when there are views, it is not safe to write directly to sstables and *also* send view updates to the paired view replica. One of the problems I can think of with concurrent updates (in parallel with the streaming, someone is modifying the same partitions/rows). For concurrent updates, as you surely remember, we need to use a row/partition lock for the RMW. If we write to sstables but not read from them yet, or if we don't properly lock rows for the right duration, we can end up breaking concurrent updates. Unless I'm missing something? So I think when views are involved, we need to pass throught the normal write code - writing to memtables, sstables, views, RMW, lock - the works - and can't optimize it. Again, unless I'm missing something. Maybe there's another way, like keeping all the rows we write to the sstable locked, but that doesn't look pretty either... If I remember correctly, Cassandra does this as well - when there are no views, it streams sstables directly (as files), but when there are views, it streams mutations and applies them using the full, normal, write path (at least, that's what I vaguely remember, I don't have the code in front of me right now).

We don't need to perform a read-before-write when invoking the view write path for a stream reason other than repair (for the same reason we don't need read-before-write for view building). We can also avoid the read-before-write if the view doesn't contain a base regular column in its key.

For repair, we could just invoke the normal view write path, which does the locking and the read-before-write. This would deal with concurrent writes. The issue, however, is that new writes wouldn't see the repaired row, in case they need to delete it. We can deal with this by invoking the full base write path or by making the sstable available for view read-before-writes.

**Comment by nyh** [comment]
On Thu, Sep 13, 2018 at 12:12 AM, Duarte Nunes <notifications@github.com>
wrote:

> I think when there are views, it is not safe to write directly to sstables
> and *also* send view updates to the paired view replica. One of the
> problems I can think of with concurrent updates (in parallel with the
> streaming, someone is modifying the same partitions/rows). For concurrent
> updates, as you surely remember, we need to use a row/partition lock for
> the RMW. If we write to sstables but not read from them yet, or if we don't
> properly lock rows for the right duration, we can end up breaking
> concurrent updates. Unless I'm missing something? So I think when views are
> involved, we need to pass throught the normal write code - writing to
> memtables, sstables, views, RMW, lock - the works - and can't optimize it.
> Again, unless I'm missing something. Maybe there's another way, like
> keeping all the rows we write to the sstable locked, but that doesn't look
> pretty either... If I remember correctly, Cassandra does this as well -
> when there are no views, it streams sstables directly (as files), but when
> there are views, it streams mutations and applies them using the full,
> normal, write path (at least, that's what I vaguely remember, I don't have
> the code in front of me right now).
>
> We don't need to perform a read-before-write when invoking the view write
> path for a stream reason other than repair
>
Ok, but my comment was on repair (again goes back to the problem that this
issue is about several issues).


> (for the same reason we don't need read-before-write for view building).
> We can also avoid the read-before-write if the view doesn't contain a base
> regular column in its key.
>

Yes, see also #3594. But we don't have this now.

> For repair, we could just invoke the normal view write path, which does
> the locking and the read-before-write.
>
Yes, that's what I suggested (I guess only when the table has views). I
thought this is also what we had originally in the code, btw.


> This would deal with concurrent writes. The issue, however, is that new
> writes wouldn't see the repaired row, in case they need to delete it.
>
We can deal with this by invoking the full base write path or by making the
> sstable available for view read-before-writes.
>

Yes, I thought the "normal write path" really means the normal write path,
including memtables, locks and everything.
I'm not sure we have a better option :-(
Repairing a table with a view will be so slow anyway (because of the extra
view updates) that I don't think the extra memtable layer will be the
problem.
